test/integration/discovery.go

```diff
@@ -147,6 +153,124 @@ var _ = Describe("Discover Cluster", func() {
 		})
 	})
 
+	Describe("Discovering affects of node with unknown state", func() {
+		var entities []*proto.EntityDTO = nil
+		var groups []*proto.GroupDTO = nil
+		var deps *appsv1.Deployment = nil
+		testName := "discovery-integration-test"
+		nodeName := "kind-worker3"
+
+		BeforeEach(func() {
+			if f.Kubeconfig.CurrentContext != "kind-kind" {
+				Skip("Skipping cluster that isn't using context kind-kind")
+			}
+
+			// create a pod to test and attach to the node soon to be in a NotReady state
+			if deps == nil {
+				var err error = nil
+				deps, err = createDeployResource(kubeClient, depSingleContainerWithResources(namespace, "", 1, false, false, false, nodeName))
+				framework.ExpectNoError(err, "Error creating test resources")
+			}
+			// stop running node worker to simulate node in NotReady state
+			execute("docker", "stop", nodeName)
+			if err := wait.PollImmediate(framework.PollInterval, framework.DefaultSingleCallTimeout, func() (bool, error) {
+				node, errInternal := kubeClient.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+				if errInternal != nil {
+					glog.Errorf("Unexpected error while finding Node %s: %v", nodeName, errInternal)
+					return false, nil
+				}
+
+				return !util.NodeIsReady(node), nil
+			}); err != nil {
+				framework.Failf("Failed to put node in a NotReady state. %v", err)
+			}
+
+			entityDTOs, groupDTOs, err := discoveryClient.DiscoverWithNewFramework(testName)
+			if err != nil {
+				framework.Failf(err.Error())
+			}
+
+			entities = entityDTOs
+			groups = groupDTOs
+		})
+
+		AfterEach(func() {
+			// restart node once finished with the test
+			execute("docker", "start", nodeName)
+			if err := wait.PollImmediate(framework.PollInterval, framework.TestContext.SingleCallTimeout, func() (bool, error) {
+				node, errInternal := kubeClient.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+				if errInternal != nil {
+					glog.Errorf("Unexpected error while finding Node %s: %v", nodeName, errInternal)
+					return false, nil
+				}
+				return util.NodeIsReady(node), nil
+			}); err != nil {
+				framework.Failf("Failed to put node in a Ready state. %v", err)
+			}
+
+		})
+
+		It("unknown node should be detected", func() {
```

Comment:
This is unnecessarily making the whole spec slower.
As far as I understand `BeforeEach()` and `AfterEach()` in the `Describe()` block is executed for each `It()` block, thus effectively starting and stopping the node 3 times over.
I think you can wrap all the three `It()` blocks in a single `Describe()` to make this faster and less error prone.